File: sdk/go/agent/verification.go
package agent

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// LocalVerifier verifies incoming requests locally using cached policies,
// revocation lists, registered DIDs, and the admin's Ed25519 public key.
// Periodically refreshes caches from the control plane.
type LocalVerifier struct {
	agentFieldURL   string
	refreshInterval time.Duration
	timestampWindow int64
	apiKey          string

	mu              sync.RWMutex
	policies        []PolicyEntry
	revokedDIDs     map[string]struct{}
	registeredDIDs  map[string]struct{}
	adminPublicKey  ed25519.PublicKey
	issuerDID       string
	lastRefresh     time.Time
	initialized     bool
}

// PolicyEntry represents a cached access policy for local evaluation.
type PolicyEntry struct {
	Name           string                       `json:"name"`
	CallerTags     []string                     `json:"caller_tags"`
	TargetTags     []string                     `json:"target_tags"`
	AllowFunctions []string                     `json:"allow_functions"`
	DenyFunctions  []string                     `json:"deny_functions"`
	Constraints    map[string]ConstraintEntry    `json:"constraints"`
	Action         string                       `json:"action"`
	Priority       int                          `json:"priority"`
	Enabled        *bool                        `json:"enabled"`
}

// ConstraintEntry represents a parameter constraint in a policy.
type ConstraintEntry struct {
	Operator string  `json:"operator"`
	Value    float64 `json:"value"`
}

// NewLocalVerifier creates a new local verifier.
func NewLocalVerifier(agentFieldURL string, refreshInterval time.Duration, apiKey string) *LocalVerifier {
	return &LocalVerifier{
		agentFieldURL:   strings.TrimRight(agentFieldURL, "/"),
		refreshInterval: refreshInterval,
		timestampWindow: 300,
		apiKey:          apiKey,
		revokedDIDs:     make(map[string]struct{}),
		registeredDIDs:  make(map[string]struct{}),
	}
}

// Refresh fetches policies, revocations, registered DIDs, and admin public key from the control plane.
func (v *LocalVerifier) Refresh() error {
	client := &http.Client{Timeout: 10 * time.Second}

	// Fetch policies
	policies, err := v.fetchPolicies(client)
	if err != nil {
		return fmt.Errorf("fetch policies: %w", err)
	}

	// Fetch revocations
	revoked, err := v.fetchRevocations(client)
	if err != nil {
		return fmt.Errorf("fetch revocations: %w", err)
	}

	// Fetch registered DIDs
	registered, err := v.fetchRegisteredDIDs(client)
	if err != nil {
		return fmt.Errorf("fetch registered DIDs: %w", err)
	}

	// Fetch admin public key
	pubKey, issuerDID, err := v.fetchAdminPublicKey(client)
	if err != nil {
		return fmt.Errorf("fetch admin public key: %w", err)
	}

	v.mu.Lock()
	defer v.mu.Unlock()
	v.policies = policies
	v.revokedDIDs = revoked
	v.registeredDIDs = registered
	v.adminPublicKey = pubKey
	v.issuerDID = issuerDID
	v.lastRefresh = time.Now()
	v.initialized = true
	return nil
}

// NeedsRefresh returns true if the cache is stale.
func (v *LocalVerifier) NeedsRefresh() bool {
	v.mu.RLock()
	defer v.mu.RUnlock()
	return time.Since(v.lastRefresh) > v.refreshInterval
}

// CheckRevocation returns true if the DID is revoked.
func (v *LocalVerifier) CheckRevocation(callerDID string) bool {
	v.mu.RLock()
	defer v.mu.RUnlock()
	_, revoked := v.revokedDIDs[callerDID]
	return revoked
}

// CheckRegistration returns true if the caller DID is registered with the control plane.
// When the cache is empty (not yet loaded), returns true to avoid blocking requests
// before the first refresh completes.
func (v *LocalVerifier) CheckRegistration(callerDID string) bool {
	v.mu.RLock()
	defer v.mu.RUnlock()
	if len(v.registeredDIDs) == 0 {
		return true // Cache not populated yet — allow
	}
	_, registered := v.registeredDIDs[callerDID]
	return registered
}

// resolvePublicKey resolves the public key bytes from a DID.
// For did:key, the public key is self-contained in the identifier:
//
//	did:key:z<base64url(0xed01 + 32-byte-pubkey)>
//
// For other DID methods, falls back to the admin public key.
func (v *LocalVerifier) resolvePublicKey(callerDID string) ed25519.PublicKey {
	if strings.HasPrefix(callerDID, "did:key:z") {
		encoded := callerDID[len("did:key:z"):]
		decoded, err := base64.RawURLEncoding.DecodeString(encoded)
		if err != nil {
			return nil
		}
		// Verify Ed25519 multicodec prefix: 0xed, 0x01
		if len(decoded) >= 34 && decoded[0] == 0xed && decoded[1] == 0x01 {
			return ed25519.PublicKey(decoded[2:34])
		}
		return nil
	}

	// Fallback: use admin public key for non-did:key methods
	v.mu.RLock()
	defer v.mu.RUnlock()
	return v.adminPublicKey
}

// VerifySignature verifies an Ed25519 DID signature on an incoming request.
// Resolves the caller's public key from their DID (did:key embeds the key
// directly; other methods fall back to the admin public key).
func (v *LocalVerifier) VerifySignature(callerDID, signatureB64, timestamp string, body []byte, nonce string) bool {
	// Validate timestamp window
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return false
	}
	now := time.Now().Unix()
	if abs64(now-ts) > v.timestampWindow {
		return false
	}

	// Resolve public key from the caller's DID
	pubKey := v.resolvePublicKey(callerDID)
	if len(pubKey) == 0 {
		return false
	}

	// Decode signature
	sigBytes, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return false
	}

	// Reconstruct the signed payload: "{timestamp}[:{nonce}]:{sha256(body)}"
	// Must match the format used by SDK signing (DIDAuthenticator)
	bodyHash := sha256.Sum256(body)
	var payload string
	if nonce != "" {
		payload = fmt.Sprintf("%s:%s:%x", timestamp, nonce, bodyHash)
	} else {
		payload = fmt.Sprintf("%s:%x", timestamp, bodyHash)
	}

	return ed25519.Verify(pubKey, []byte(payload), sigBytes)
}

func (v *LocalVerifier) doRequest(client *http.Client, path string) (*http.Response, error) {
	req, err := http.NewRequest(http.MethodGet, v.agentFieldURL+path, nil)
	if err != nil {
		return nil, err
	}
	if v.apiKey != "" {
		req.Header.Set("X-API-Key", v.apiKey)
	}
	return client.Do(req)
}

func (v *LocalVerifier) fetchPolicies(client *http.Client) ([]PolicyEntry, error) {
	resp, err := v.doRequest(client, "/api/v1/policies")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	var result struct {
		Policies []PolicyEntry `json:"policies"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Policies, nil
}

func (v *LocalVerifier) fetchRevocations(client *http.Client) (map[string]struct{}, error) {
	resp, err := v.doRequest(client, "/api/v1/revocations")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	var result struct {
		RevokedDIDs []string `json:"revoked_dids"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	revoked := make(map[string]struct{}, len(result.RevokedDIDs))
	for _, d := range result.RevokedDIDs {
		revoked[d] = struct{}{}
	}
	return revoked, nil
}

func (v *LocalVerifier) fetchRegisteredDIDs(client *http.Client) (map[string]struct{}, error) {
	resp, err := v.doRequest(client, "/api/v1/registered-dids")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	var result struct {
		RegisteredDIDs []string `json:"registered_dids"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	registered := make(map[string]struct{}, len(result.RegisteredDIDs))
	for _, d := range result.RegisteredDIDs {
		registered[d] = struct{}{}
	}
	return registered, nil
}

func (v *LocalVerifier) fetchAdminPublicKey(client *http.Client) (ed25519.PublicKey, string, error) {
	resp, err := v.doRequest(client, "/api/v1/admin/public-key")
	if err != nil {
		return nil, "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	var result struct {
		IssuerDID    string                 `json:"issuer_did"`
		PublicKeyJWK map[string]interface{} `json:"public_key_jwk"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, "", err
	}

	// Parse Ed25519 public key from JWK
	xValue, ok := result.PublicKeyJWK["x"].(string)
	if !ok {
		return nil, "", fmt.Errorf("missing 'x' in public key JWK")
	}
	pubKeyBytes, err := base64.RawURLEncoding.DecodeString(xValue)
	if err != nil {
		return nil, "", fmt.Errorf("decode public key: %w", err)
	}
	if len(pubKeyBytes) != ed25519.PublicKeySize {
		return nil, "", fmt.Errorf("invalid public key size: %d", len(pubKeyBytes))
	}
	return ed25519.PublicKey(pubKeyBytes), result.IssuerDID, nil
}

func abs64(x int64) int64 {
	if x < 0 {
		neg := -x
		if neg < 0 {
			// Overflow: -math.MinInt64 overflows back to negative.
			return math.MaxInt64
		}
		return neg
	}
	return x
}

// EvaluatePolicy evaluates access policies locally.
func (v *LocalVerifier) EvaluatePolicy(callerTags, targetTags []string, functionName string, inputParams map[string]any) bool {
	v.mu.RLock()
	policies := make([]PolicyEntry, len(v.policies))
	copy(policies, v.policies)
	v.mu.RUnlock()

	if len(policies) == 0 {
		return false // No policies — fail closed
	}

	// Sort by priority descending so highest-priority policies are evaluated first.
	sort.Slice(policies, func(i, j int) bool {
		return policies[i].Priority > policies[j].Priority
	})

	for _, policy := range policies {
		if policy.Enabled != nil && !*policy.Enabled {
			continue
		}

		// Check caller tags match
		if len(policy.CallerTags) > 0 && !anyTagMatch(callerTags, policy.CallerTags) {
			continue
		}

		// Check target tags match
		if len(policy.TargetTags) > 0 && !anyTagMatch(targetTags, policy.TargetTags) {
			continue
		}

		// Check deny functions first
		if len(policy.DenyFunctions) > 0 && functionMatches(functionName, policy.DenyFunctions) {
			return false
		}

		// Check allow functions
		if len(policy.AllowFunctions) > 0 && !functionMatches(functionName, policy.AllowFunctions) {
			continue
		}

		// Check constraints
		if len(policy.Constraints) > 0 && inputParams != nil {
			if !evaluateConstraints(policy.Constraints, functionName, inputParams) {
				return false
			}
		}

		action := policy.Action
		if action == "" {
			action = "allow"
		}
		return action == "allow"
	}

	return true // No matching policy — allow by default
}

func anyTagMatch(have, want []string) bool {
	for _, w := range want {
		for _, h := range have {
			if h == w {
				return true
			}
		}
	}
	return false
}

func functionMatches(name string, patterns []string) bool {
	for _, p := range patterns {
		if matchWildcard(name, p) {
			return true
		}
	}
	return false
}

func matchWildcard(name, pattern string) bool {
	if pattern == "*" {
		return true
	}
	if strings.HasSuffix(pattern, "*") {
		return strings.HasPrefix(name, strings.TrimSuffix(pattern, "*"))
	}
	if strings.HasPrefix(pattern, "*") {
		return strings.HasSuffix(name, strings.TrimPrefix(pattern, "*"))
	}
	return name == pattern
}

func evaluateConstraints(constraints map[string]ConstraintEntry, functionName string, inputParams map[string]any) bool {
	for paramName, constraint := range constraints {
		val, ok := inputParams[paramName]
		if !ok {
			return false // Fail closed: constrained parameter missing from input
		}
		numVal, err := toFloat64(val)
		if err != nil {
			return false // Fail closed: cannot convert constrained parameter to numeric
		}
		switch constraint.Operator {
		case "<=":
			if numVal > constraint.Value {
				return false
			}
		case ">=":
			if numVal < constraint.Value {
				return false
			}
		case "<":
			if numVal >= constraint.Value {
				return false
			}
		case ">":
			if numVal <= constraint.Value {
				return false
			}
		case "==":
			if math.Abs(numVal-constraint.Value) > 1e-9 {
				return false
			}
		}
	}
	return true
}

func toFloat64(v any) (float64, error) {
	switch val := v.(type) {
	case float64:
		return val, nil
	case float32:
		return float64(val), nil
	case int:
		return float64(val), nil
	case int64:
		return float64(val), nil
	case json.Number:
		return val.Float64()
	case string:
		return strconv.ParseFloat(val, 64)
	default:
		return 0, fmt.Errorf("unsupported type %T", v)
	}
}

File: sdk/python/agentfield/verification.py
"""
Local verification for AgentField SDK.

Provides decentralized verification of incoming requests by caching policies,
revocation lists, and the admin's public key from the control plane. Agents
can verify DID signatures and evaluate access policies locally without
hitting the control plane for every call.
"""

import base64
import hashlib
import time
from typing import Any, Dict, List, Optional, Set

from .logger import get_logger

logger = get_logger(__name__)

# DID auth headers (same as did_auth.py)
HEADER_CALLER_DID = "X-Caller-DID"
HEADER_DID_SIGNATURE = "X-DID-Signature"
HEADER_DID_TIMESTAMP = "X-DID-Timestamp"


class LocalVerifier:
    """
    Verifies incoming requests locally using cached policies, revocations,
    and the admin's Ed25519 public key.

    Periodically refreshes caches from the control plane. If the control plane
    is unreachable, continues using stale caches until TTL expires.
    """

    def __init__(
        self,
        agentfield_url: str,
        refresh_interval: int = 300,
        timestamp_window: int = 300,
        api_key: Optional[str] = None,
    ):
        """
        Initialize the local verifier.

        Args:
            agentfield_url: Base URL of the AgentField control plane
            refresh_interval: Seconds between cache refreshes (default: 300 = 5 min)
            timestamp_window: Allowed timestamp skew in seconds (default: 300 = 5 min)
            api_key: Optional API key for authenticating with the control plane
        """
        self.agentfield_url = agentfield_url.rstrip("/")
        self.refresh_interval = refresh_interval
        self.timestamp_window = timestamp_window
        self.api_key = api_key

        # Cached data
        self.policies: List[Dict[str, Any]] = []
        self.revoked_dids: Set[str] = set()
        self.registered_dids: Set[str] = set()
        self.admin_public_key_jwk: Optional[Dict[str, Any]] = None
        self.issuer_did: Optional[str] = None

        # Cache metadata
        self._last_refresh: float = 0
        self._initialized: bool = False

    async def refresh(self) -> bool:
        """
        Fetch policies, revocations, and admin public key from the control plane.

        Returns:
            True if refresh succeeded, False otherwise (stale cache still used)
        """
        try:
            import aiohttp
        except ImportError:
            logger.warning("aiohttp not available, cannot refresh verification cache")
            return False

        headers = {}
        if self.api_key:
            headers["X-API-Key"] = self.api_key

        success = True
        async with aiohttp.ClientSession() as session:
            # Fetch policies
            try:
                async with session.get(
                    f"{self.agentfield_url}/api/v1/policies",
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=10),
                ) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        self.policies = data.get("policies", []) or []
                        logger.debug(f"Refreshed {len(self.policies)} policies")
                    else:
                        logger.warning(f"Failed to fetch policies: HTTP {resp.status}")
                        success = False
            except Exception as e:
                logger.warning(f"Failed to fetch policies: {e}")
                success = False

            # Fetch revocations
            try:
                async with session.get(
                    f"{self.agentfield_url}/api/v1/revocations",
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=10),
                ) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        self.revoked_dids = set(data.get("revoked_dids", []))
                        logger.debug(f"Refreshed {len(self.revoked_dids)} revoked DIDs")
                    else:
                        logger.warning(f"Failed to fetch revocations: HTTP {resp.status}")
                        success = False
            except Exception as e:
                logger.warning(f"Failed to fetch revocations: {e}")
                success = False

            # Fetch registered DIDs
            try:
                async with session.get(
                    f"{self.agentfield_url}/api/v1/registered-dids",
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=10),
                ) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        self.registered_dids = set(data.get("registered_dids", []))
                        logger.debug(f"Refreshed {len(self.registered_dids)} registered DIDs")
                    else:
                        logger.warning(f"Failed to fetch registered DIDs: HTTP {resp.status}")
                        success = False
            except Exception as e:
                logger.warning(f"Failed to fetch registered DIDs: {e}")
                success = False

            # Fetch admin public key
            try:
                async with session.get(
                    f"{self.agentfield_url}/api/v1/admin/public-key",
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=10),
                ) as resp:
                    if resp.status == 200:
                        data = await resp.json()
                        self.admin_public_key_jwk = data.get("public_key_jwk")
                        self.issuer_did = data.get("issuer_did")
                        logger.debug(f"Refreshed admin public key (issuer: {self.issuer_did})")
                    else:
                        logger.warning(f"Failed to fetch admin public key: HTTP {resp.status}")
                        success = False
            except Exception as e:
                logger.warning(f"Failed to fetch admin public key: {e}")
                success = False

        if success:
            self._last_refresh = time.time()
            self._initialized = True

        return success

    @property
    def needs_refresh(self) -> bool:
        """Check if the cache is stale and needs refreshing."""
        if not self._initialized:
            return True
        return time.time() - self._last_refresh > self.refresh_interval

    def check_revocation(self, caller_did: str) -> bool:
        """
        Check if a caller DID is in the revocation list.

        Args:
            caller_did: The DID to check

        Returns:
            True if revoked, False if not revoked
        """
        return caller_did in self.revoked_dids

    def check_registration(self, caller_did: str) -> bool:
        """
        Check if a caller DID is registered with the control plane.

        Returns True if registered (known), False if unknown. When the
        registered DIDs cache is empty (not yet loaded), returns True to
        avoid blocking requests before the first refresh completes.
        """
        if not self.registered_dids:
            # Cache not yet populated — allow to avoid blocking before first refresh.
            return True
        return caller_did in self.registered_dids

    def verify_signature(
        self,
        caller_did: str,
        signature_b64: str,
        timestamp: str,
        body: bytes,
        nonce: str = "",
    ) -> bool:
        """
        Verify an Ed25519 DID signature on an incoming request.

        Resolves the caller's public key from their DID (did:key embeds the key
        directly; other methods fall back to the admin public key).

        Args:
            caller_did: Caller's DID identifier
            signature_b64: Base64-encoded Ed25519 signature
            timestamp: Unix timestamp string from the request
            body: Request body bytes
            nonce: Optional nonce from X-DID-Nonce header

        Returns:
            True if signature is valid, False otherwise
        """
        # Validate timestamp window
        try:
            ts = int(timestamp)
            now = int(time.time())
            if abs(now - ts) > self.timestamp_window:
                logger.debug(f"Timestamp expired: {now - ts}s drift (window: {self.timestamp_window}s)")
                return False
        except (ValueError, TypeError):
            logger.debug("Invalid timestamp format")
            return False

        try:
            from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        except ImportError:
            logger.warning("cryptography library not available for signature verification")
            return False

        try:
            # Resolve public key from the caller's DID
            public_key_bytes = self._resolve_public_key(caller_did)
            if public_key_bytes is None:
                logger.debug(f"Could not resolve public key for DID: {caller_did}")
                return False
            public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes)

            # Reconstruct the signed payload: "{timestamp}[:{nonce}]:{sha256(body)}"
            # Must match the format used by SDK signing (did_auth.py)
            body_hash = hashlib.sha256(body).hexdigest()
            if nonce:
                payload = f"{timestamp}:{nonce}:{body_hash}".encode("utf-8")
            else:
                payload = f"{timestamp}:{body_hash}".encode("utf-8")

            # Decode the signature
            signature_bytes = base64.b64decode(signature_b64)

            # Verify
            public_key.verify(signature_bytes, payload)
            return True

        except Exception as e:
            logger.debug(f"Signature verification failed: {e}")
            return False

    def _resolve_public_key(self, caller_did: str) -> Optional[bytes]:
        """
        Resolve the public key bytes from a DID.

        For did:key, the public key is self-contained in the identifier:
          did:key:z<base64url(0xed01 + 32-byte-pubkey)>

        For other DID methods, falls back to the admin public key.
        """
        if caller_did.startswith("did:key:z"):
            try:
                encoded = caller_did[len("did:key:z"):]
                decoded = base64.urlsafe_b64decode(encoded + "==")
                # Verify Ed25519 multicodec prefix: 0xed, 0x01
                if len(decoded) >= 34 and decoded[0] == 0xED and decoded[1] == 0x01:
                    return decoded[2:34]
                logger.debug(f"Invalid multicodec prefix in did:key: {decoded[:2].hex()}")
                return None
            except Exception as e:
                logger.debug(f"Failed to decode did:key public key: {e}")
                return None

        # Fallback: use admin public key for non-did:key methods
        if self.admin_public_key_jwk:
            try:
                x_value = self.admin_public_key_jwk.get("x", "")
                padding = 4 - (len(x_value) % 4)
                if padding != 4:
                    x_value += "=" * padding
                return base64.urlsafe_b64decode(x_value)
            except Exception as e:
                logger.debug(f"Failed to decode admin public key: {e}")
                return None

        logger.debug("No public key available for verification")
        return None

    def evaluate_policy(
        self,
        caller_tags: List[str],
        target_tags: List[str],
        function_name: str,
        input_params: Optional[Dict[str, Any]] = None,
    ) -> bool:
        """
        Evaluate access policies locally.

        Finds matching policies based on caller/target tags and function name,
        then evaluates constraints.

        Args:
            caller_tags: Tags associated with the calling agent
            target_tags: Tags associated with the target agent
            function_name: Name of the function being called
            input_params: Input parameters for constraint evaluation

        Returns:
            True if access is allowed, False if denied
        """
        if not self.policies:
            # Fail closed: no policies loaded means we cannot verify access.
            # This prevents bypassing authorization when policies fail to load.
            return False

        # Sort policies by priority (descending)
        sorted_policies = sorted(
            self.policies,
            key=lambda p: p.get("priority", 0),
            reverse=True,
        )

        for policy in sorted_policies:
            if not policy.get("enabled", True):
                continue

            # Check if caller tags match
            policy_caller_tags = policy.get("caller_tags", [])
            if policy_caller_tags and not any(t in caller_tags for t in policy_caller_tags):
                continue

            # Check if target tags match
            policy_target_tags = policy.get("target_tags", [])
            if policy_target_tags and not any(t in target_tags for t in policy_target_tags):
                continue

            # Check function allow/deny lists
            allow_functions = policy.get("allow_functions", [])
            deny_functions = policy.get("deny_functions", [])

            # Check deny list first
            if deny_functions and _function_matches(function_name, deny_functions):
                return False

            # Check allow list
            if allow_functions and not _function_matches(function_name, allow_functions):
                continue

            # Check constraints
            constraints = policy.get("constraints", {})
            if constraints and input_params:
                if not _evaluate_constraints(constraints, function_name, input_params):
                    return False

            # Policy action
            action = policy.get("action", "allow")
            return action == "allow"

        # No matching policy found — allow by default
        return True


def _function_matches(function_name: str, patterns: List[str]) -> bool:
    """Check if a function name matches any of the patterns (supports * wildcards)."""
    import fnmatch

    for pattern in patterns:
        if fnmatch.fnmatch(function_name, pattern):
            return True
    return False


def _evaluate_constraints(
    constraints: Dict[str, Any],
    function_name: str,
    input_params: Dict[str, Any],
) -> bool:
    """Evaluate parameter constraints for a function call."""
    # Constraints can be keyed by function name or parameter name
    func_constraints = constraints.get(function_name, constraints)
    if not isinstance(func_constraints, dict):
        return True

    for param_name, constraint in func_constraints.items():
        if param_name not in input_params:
            continue

        value = input_params[param_name]
        if isinstance(constraint, dict):
            operator = constraint.get("operator", "")
            threshold = constraint.get("value")
            if threshold is None:
                continue

            try:
                value = float(value)
                threshold = float(threshold)
            except (ValueError, TypeError):
                # Fail closed: invalid constraint values should deny access
                # rather than silently skipping the constraint check
                return False

            if operator == "<=" and value > threshold:
                return False
            elif operator == ">=" and value < threshold:
                return False
            elif operator == "<" and value >= threshold:
                return False
            elif operator == ">" and value <= threshold:
                return False
            elif operator == "==" and value != threshold:
                return False

    return True
